libr/bin/mangling/swift-sd.c

```diff
@@ -171,33 +177,44 @@ static char *swift_demangle_lib(const char *s) {
 	return NULL;
 }
 
+static const char *str_seek (const char *s, int n) {
+	int i;
+	for (i = 0; i < n && *s != 0; i++) {
+		s++;
+	}
+	return s;
+}
+
+static const char *str_removeprefix (const char *s, const char *prefix) {
+	if (r_str_startswith (s, prefix)) {
+		s = s + strlen (prefix);
+	}
+	return s;
+}
+
+static void demangle_alloc_fail(RStrBuf *out) {
+	r_assert_log (R_LOGLVL_WARN, __FILE__, "demangle malloc failed");
+	r_strbuf_free (out);
+}
+
 R_API char *r_bin_demangle_swift(const char *s, bool syscmd, bool trylib) {
-#define STRCAT_BOUNDS(x) if (((x) + 2 + strlen (out)) > sizeof (out)) break;
-	char out[1024];
 	int i, len, is_generic = 0;
 	int is_first = 1;
 	int is_last = 0;
 	int retmode = 0;
-	if (!strncmp (s, "imp.", 4)) {
-		s = s + 4;
-	}
-	if (!strncmp (s, "reloc.", 6)) {
-		s = s + 6;
-	}
-
-	if (!strncmp (s, "__", 2)) {
-		s = s + 2;
-	}
+	s = str_removeprefix (s, "imp.");
+	s = str_removeprefix (s, "reloc.");
+	s = str_removeprefix (s, "__");
 	char *res = NULL;
 	if (trylib) {
 		res = swift_demangle_lib (s);
 		if (res) {
 			return res;
 		}
 	}
-	if (*s != 'T' && strncmp (s, "_T", 2) && strncmp (s, "__T", 3)) {
+	if (*s != 'T' && !r_str_startswith (s, "_T") && !r_str_startswith (s, "__T")) {
 		// modern swift symbols not yet supported in this parser (only via trylib)
-		if (strncmp (s, "$s", 2)) {
+		if (!r_str_startswith (s, "$s")) {
```

Comment:
These 3 lines make no sense to me

Reply:
why not? if the string doesnt starts with '$s' (and its clearly not starting with `T` or _T, then its not a swift string. am i missing anything?

anyway, i wrote this code before swift was opensource, so now it would be better to rewrite it and support more constrictions. we cant reuse the ofificla implementation because its written in swift, and iirc there's another one written in c++. but for mac/ios users. r2 uses the system implementation for this which covers much more cases

Reply:
ahh thanks makes sense now, thanks for the clarification 👍🏻 I was confused because usually `r_str_startswith` checks aren't negated